--- annobin.orig/plugin/x86_64.annobin.cc	2018-11-05 16:31:45.253977255 +0000
+++ annobin-8.23/plugin/x86_64.annobin.cc	2018-11-05 16:35:40.572266029 +0000
@@ -239,7 +239,10 @@ annobin_target_specific_loader_notes (vo
   annobin_inform (1, "Creating notes for the dynamic loader");
 
   fprintf (asm_out_file, "\t.section %s, \"a\", %%note\n", NOTE_GNU_PROPERTY_SECTION_NAME);
-  fprintf (asm_out_file, "\t.balign 4\n");
+  if (annobin_is_64bit)
+    fprintf (asm_out_file, "\t.balign 8\n");
+  else
+    fprintf (asm_out_file, "\t.balign 4\n");
 
   ptr = buffer;
 
