diff -rup annobin.orig/plugin/annobin.cc annobin-5.7/plugin/annobin.cc
--- annobin.orig/plugin/annobin.cc	2018-10-15 11:57:59.263140156 +0100
+++ annobin-5.7/plugin/annobin.cc	2018-10-15 12:00:17.358170612 +0100
@@ -1158,10 +1158,14 @@ plugin_init (struct plugin_name_args *
     {
       bool fail = false;
 
-      if (strcmp (version->basever, gcc_version.basever))
+      /* plugin_default_version_check is very strict and requires that the
+	 major, minor and revision numbers all match.  Since annobin only
+	 lightly touches gcc we assume that major number compatibility will
+	 be sufficient...  */
+      if (strncmp (version->basever, gcc_version.basever, strchr (version->basever, '.') - version->basever))
 	{
 	  annobin_inform (0, _("Error: plugin built for compiler version (%s) but run with compiler version (%s)"),
-			  version->basever, gcc_version.basever);
+			  gcc_version.basever, version->basever);
 	  fail = true;
 	}
 
